logstash-core/lib/logstash/plugins/registry.rb

```diff
@@ -1,86 +1,243 @@
     def initialize
-      @registry = {}
-      @logger = self.logger
+     @registry = {}
+     @hooks = HooksRegistry.new
     end
 
-    def lookup(type, plugin_name, &block)
+    def setup!
+      load_available_plugins
+      execute_universal_plugins
+    end
 
-      plugin = Plugin.new(type, plugin_name)
+    def execute_universal_plugins
+      @registry.values
+        .select { |specification| specification.is_a?(UniversalPluginSpecification) }
+        .each { |specification| specification.register(hooks, LogStash::SETTINGS) }
+    end
 
-      if plugin.installed?
-        return @registry[plugin.path] if registered?(plugin.path)
-        require plugin.path
-        klass = @registry[plugin.path]
-        if block_given? # if provided pass a block to do validation
-          raise LoadError unless block.call(klass, plugin_name)
+    def load_available_plugins
+      GemRegistry.logstash_plugins.each do |plugin_context|
+        # When a plugin has a HOOK_FILE defined, its the responsability of the plugin
+        # to register itself to the registry of available plugins.
+        #
+        # Legacy plugin will lazy register themselves
+        if plugin_context.has_hooks?
+          begin
+            logger.debug("Executing hooks", :name => plugin_context.name, :type => plugin_context.type, :hooks_file => plugin_context.hooks_file)
+            plugin_context.execute_hooks!
+          rescue => e
+            logger.error("error occured when loading plugins hooks file", :name => plugin_context.name, :type => plugin_context.type, :exception => e.message, :stacktrace => e.backtrace )
+          end
         end
-        return klass
-      else
-        # The plugin was defined directly in the code, so there is no need to use the
-        # require way of loading classes
-        return @registry[plugin.path] if registered?(plugin.path)
-        raise LoadError
       end
-    rescue => e
-      @logger.warn("Problems loading a plugin with", :type => type, :name => plugin, :path => plugin.path,
-                   :error_message => e.message, :error_class => e.class, :error_backtrace => e.backtrace)
-      raise LoadError, "Problems loading the requested plugin named #{plugin_name} of type #{type}. Error: #{e.class} #{e.message}"
     end
 
-    def register(path, klass)
-      @registry[path] = klass
+    def lookup(type, plugin_name, &block)
+      plugin = get(type, plugin_name)
+      # Assume that we have a legacy plugin
+      if plugin.nil?
+        begin
+          path = "logstash/#{type}s/#{plugin_name}"
+
+          begin
+            require path
+          rescue LoadError
+            # Plugin might be already defined in the current scope
+            # This scenario often happen in test when we write an adhoc class
+          end
+
+          klass = namespace_lookup(type, plugin_name)
+          plugin = lazy_add(type, plugin_name, klass)
+        rescue => e
+          logger.warn("Problems loading a plugin with",
```

Comment:
Not all the type, if the namespace_lookup fails to find the name of the plugin class it will catched by this rescue.